src/analysis_and_optimization/Mir_utils.ml

```diff
@@ -367,17 +375,22 @@ and idx_depth i =
   | Between (e1, e2) -> max (expr_depth e1) (expr_depth e2)
 
 let ad_level_sup l =
-  if List.exists l ~f:(fun x -> Expr.Typed.adlevel_of x = AutoDiffable) then
-    UnsizedType.AutoDiffable
+  if
+    List.exists l ~f:(fun x ->
+        UnsizedType.is_autodifftype @@ Expr.Typed.adlevel_of x )
+  then UnsizedType.AutoDiffable
   else DataOnly
```

Comment:
This is better but it's not used correctly in the following code (and the name makes me think of `UnsizedType.lub_ad_type` which this is not)

* `FunApp` needs `adlevel= UnsizedType.fill_adtype_for_type (ad_level_sup l) e.meta.type_`
* `TernaryIf` needs `adlevel= UnsizedType.lub_ad_level [e1; e2; e3]`
* `EAnd` and `EOr` I _think_ could get aways with `adlevel= DataOnly` but maybe should use `lub_ad_type`
* `Promotion` is tricky. `e.meta` should have `adlevel= ad` but I think this pass might want to remove the promotion if it just changes autodiffability but it's difficult to know if that breaks anything else so probably better to just leave it here.
* `Indexed` probably should ignore adlevels in the indices and just go with `adlevel= ixed.meta.adlevel`

Since `ad_level_sup` is a small function only used in one place, I'd prefer to inline it.